src/main/java/net/snowflake/ingest/utils/IcebergDataTypeParser.java

```diff
@@ -208,4 +223,61 @@ public static Types.MapType mapFromJson(JsonNode json) {
       return Types.MapType.ofOptional(keyId, valueId, keyType, valueType);
     }
   }
+
+  private static org.apache.parquet.schema.Type replaceWithOriginalFieldName(
+      org.apache.parquet.schema.Type parquetType, Type icebergType, String fieldName) {
+    if (parquetType.isPrimitive() != icebergType.isPrimitiveType()
+        || (!parquetType.isPrimitive()
+            && parquetType.getLogicalTypeAnnotation()
+                == null /* ignore outer layer of map or list */
+            && parquetType.asGroupType().getFieldCount()
+                != icebergType.asNestedType().fields().size())) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Parquet type and Iceberg type mismatch: %s, %s", parquetType, icebergType));
+    }
+    if (parquetType.isPrimitive()) {
+      /* rename field name */
+      return org.apache.parquet.schema.Types.primitive(
+              parquetType.asPrimitiveType().getPrimitiveTypeName(), parquetType.getRepetition())
+          .as(parquetType.asPrimitiveType().getLogicalTypeAnnotation())
+          .id(parquetType.getId().intValue())
+          .length(parquetType.asPrimitiveType().getTypeLength())
+          .named(fieldName);
+    } else {
+      org.apache.parquet.schema.Types.GroupBuilder<org.apache.parquet.schema.GroupType> builder =
+          org.apache.parquet.schema.Types.buildGroup(parquetType.getRepetition());
+      for (org.apache.parquet.schema.Type parquetFieldType :
+          parquetType.asGroupType().getFields()) {
+        if (parquetFieldType.getId() == null) {
+          /* middle layer of map or list. Skip this level as parquet's using 3-level list/map while iceberg's using 2-level list/map */
+          builder.addField(
+              replaceWithOriginalFieldName(
+                  parquetFieldType, icebergType, parquetFieldType.getName()));
+        } else {
+          Types.NestedField icebergField =
+              icebergType.asNestedType().field(parquetFieldType.getId().intValue());
+          if (icebergField == null) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Cannot find Iceberg field with id %d in Iceberg type: %s",
+                    parquetFieldType.getId().intValue(), icebergType));
+          }
+          builder.addField(
+              replaceWithOriginalFieldName(
+                  parquetFieldType,
+                  icebergField.type(),
+                  icebergField.name().equals(EMPTY_FIELD_CHAR)
+                      ? ""
+                      : icebergField
+                          .name()
+                          .replace(EMPTY_FIELD_CHAR + EMPTY_FIELD_CHAR, EMPTY_FIELD_CHAR)));
+        }
+      }
+      if (parquetType.getId() != null) {
+        builder.id(parquetType.getId().intValue());
+      }
```

Comment:
else ?

Reply:
The middle layer of Iceberg list/map does not have id. We don't want to set it.